src/components/implementation/no_interface/vkernel/vkernel.c

```diff
@@ -0,0 +1,195 @@
+
+		vm_info->pt = cos_pgtbl_alloc(vk_cinfo);
+		assert(vm_info->pt);
+
+		vm_info->utpt = cos_pgtbl_alloc(vk_cinfo);
+		assert(vm_info->utpt);
+
+		vm_info->cc = cos_comp_alloc(vk_cinfo, vm_info->ct, vm_info->pt, (vaddr_t)&cos_upcall_entry);
+		assert(vm_info->cc);
+
+		cos_meminfo_init(&vm_cinfo->mi, BOOT_MEM_KM_BASE, COS_VIRT_MACH_UNTYPED_SIZE, vm_info->utpt);
+		cos_compinfo_init(vm_cinfo, vm_info->pt, vm_info->ct, vm_info->cc,
+				(vaddr_t)BOOT_MEM_VM_BASE, VM_CAPTBL_FREE, vk_cinfo);
+
+		vm_info->initthd = cos_thd_alloc(vk_cinfo, vm_cinfo->comp_cap, vm_init, (void *)id);
+		assert(vm_info->initthd);
+		vm_info->inittid = (thdid_t)cos_introspect(vk_cinfo, vm_info->initthd, 9);
+		printc("\tInit thread= cap:%x tid:%x\n", (unsigned int)vm_info->initthd, (unsigned int)vm_info->inittid);
+
+		printc("\tCopying pgtbl, captbl, component capabilities\n");
+		ret = cos_cap_cpy_at(vm_cinfo, BOOT_CAPTBL_SELF_CT, vk_cinfo, vm_info->ct);
+		assert(ret == 0);
+		ret = cos_cap_cpy_at(vm_cinfo, BOOT_CAPTBL_SELF_PT, vk_cinfo, vm_info->pt);
+		assert(ret == 0);
+		ret = cos_cap_cpy_at(vm_cinfo, BOOT_CAPTBL_SELF_UNTYPED_PT, vk_cinfo, vm_info->utpt);
+		assert(ret == 0);
+		ret = cos_cap_cpy_at(vm_cinfo, BOOT_CAPTBL_SELF_COMP, vk_cinfo, vm_info->cc);
+		assert(ret == 0);
+
+		printc("\tCreating and copying required initial capabilities\n");
+		/*
+		 * TODO: Multi-core support to create INITIAL Capabilities per core
+		 */
+		ret = cos_cap_cpy_at(vm_cinfo, BOOT_CAPTBL_SELF_INITTHD_BASE, vk_cinfo, vm_info->initthd);
+		assert(ret == 0);
+		ret = cos_cap_cpy_at(vm_cinfo, BOOT_CAPTBL_SELF_INITHW_BASE, vk_cinfo, BOOT_CAPTBL_SELF_INITHW_BASE);
+		assert(ret == 0);
+		ret = cos_cap_cpy_at(vm_cinfo, VM_CAPTBL_SELF_EXITTHD_BASE, vk_cinfo, vm_info->exitthd);
+		assert(ret == 0);
+		
+		vm_info->inittcap = cos_tcap_alloc(vk_cinfo, TCAP_PRIO_MAX);
+		assert(vm_info->inittcap);
+
+		vm_info->initrcv = cos_arcv_alloc(vk_cinfo, vm_info->initthd, vm_info->inittcap, vk_cinfo->comp_cap, BOOT_CAPTBL_SELF_INITRCV_BASE);
+		assert(vm_info->initrcv);
+
+		ret = cos_tcap_transfer(vm_info->initrcv, BOOT_CAPTBL_SELF_INITTCAP_BASE, TCAP_RES_INF, TCAP_PRIO_MAX);
+		assert(ret == 0);
+
+		ret = cos_cap_cpy_at(vm_cinfo, BOOT_CAPTBL_SELF_INITTCAP_BASE, vk_cinfo, vm_info->inittcap);
+		assert(ret == 0);
+		ret = cos_cap_cpy_at(vm_cinfo, BOOT_CAPTBL_SELF_INITRCV_BASE, vk_cinfo, vm_info->initrcv);
+		assert(ret == 0);
+
+		/*
+		 * Create send end-point in VKernel to each VM's INITRCV end-point
+		 */
+		vk_info.vminitasnd[id] = cos_asnd_alloc(vk_cinfo, vm_info->initrcv, vk_cinfo->captbl_cap);
+		assert(vk_info.vminitasnd[id]);
+
+		/*
+		 * Create and copy booter comp virtual memory to each VM
+		 */
+		vm_range = (int)cos_get_heap_ptr() - BOOT_MEM_VM_BASE;
+		assert(vm_range > 0);
+		printc("\tMapping in Booter component's virtual memory (range:%u)\n", vm_range);
+		for (i = 0; i < vm_range; i += PAGE_SIZE) {
+			vaddr_t spg = (vaddr_t)cos_page_bump_alloc(vk_cinfo), dpg;
```

Comment:
since there are no data-structures that given any hint as to what this means, src_pg seems more descriptive.